packages/core/src/verification/signature-verifier.ts

```diff
@@ -0,0 +1,544 @@
+    for (const gateway of this.trustedGateways) {
+      try {
+        const url = `${gateway.toString()}${rootTransactionId}`;
+        const response = await fetch(url, {
+          method: 'GET',
+          redirect: 'follow',
+          mode: 'cors',
+          headers: {
+            Range: `bytes=${rangeStart}-${rangeEnd}`,
+            'Cache-Control': 'no-cache',
+          },
+        });
+
+        if (!response.ok || response.body === null) {
+          continue;
+        }
+
+        // create the data item object from just the signature bytes, so we can get the components easily
+        const trustedDataItemHeaderBytes = Buffer.from(
+          await response.arrayBuffer(),
+        );
+        const dataItem = new DataItem(trustedDataItemHeaderBytes);
+
+        // first verify the data item id matches the txId before we do any other verification
+        const dataItemId = await dataItem.id;
+        if (dataItemId !== txId) {
+          throw new Error('Data item ID does not match txId', {
+            cause: { dataItemId, txId },
+          });
+        }
+
+        // now get the components needed for verification from the DataItem
+        const signatureType = dataItem.signatureType;
+        const signature = dataItem.rawSignature;
+        const owner = dataItem.rawOwner;
+        const target = dataItem.rawTarget;
+        const anchor = dataItem.rawAnchor;
+        const tags = dataItem.rawTags;
+
+        return {
+          signatureType,
+          signature,
+          owner,
+          target,
+          anchor,
+          tags,
+        };
+      } catch (error: any) {
+        this.logger.debug('Failed to fetch data item signature bytes', {
+          error: error.message,
+          stack: error.stack,
+          txId,
+        });
+        // Continue to the next gateway if there's an error
+        continue;
+      }
+    }
+
+    throw new Error('Failed to fetch data item signature bytes', {
+      cause: { txId, rootTransactionId, dataItemOffset, dataItemDataOffset },
+    });
+  }
+
+  /**
+   * Verifies the signature of a data item using trusted gateways to get the signature data
+   * and then verifying it against the data payload.
+   *
+   * This implementation follows the ANS-104 standard for verifying data item signatures:
+   * 1. Get the signature data from the trusted gateway
+   * 2. Parse the signature components (owner, signature type, tags, etc.)
+   * 3. Calculate the deep hash of all components and the data
+   * 4. Verify the signature against the deep hash
+   *
+   * @param data - The data stream to verify
+   * @param txId - The transaction ID of the data
+   */
+  async verifyData({
+    data,
+    txId,
+  }: {
+    data: DataStream;
+    txId: string;
+  }): Promise<void> {
+    // fetch signature data from trusted gateway
+    const trustedSignatureData = await this.getSignatureData({ txId });
```

Comment:
You could also fetch this from the untrusted gateway. Unless the underlying cryptography is broken, there's no way to get a matching ID unless the signature data matches what was used to generate the ID in the first place.